### ISSUE TITLE
Unsolvable constraint: `inf`

### ISSUE THREAD

**Issue by brianhuffman** [GaloisInc/cryptol#837]
This is a bit of an ill-formed error message, as it lists something of numeric kind as an "unsolvable constraint":
```
Cryptol> 1 : [inf]

[error] at <interactive>:1:1--1:2:
  Unsolvable constraints:
    • inf
        arising from
        use of literal or demoted expression
        at <interactive>:1:1--1:2
    • Reason: `inf` is not finite
```
I think it should instead say
```
  Unsolvable constraints:
    • fin inf
```

**Comment by yav** [comment]
Hm, I thought I fixed that.

Anyway, this happens because I changed how we treat the special "error" type.  Previously if during simplification of types we encountered an error (e.g., due to a partial type function), we would just produce a dummy type, something like `Error "Description"`, and then these errors would sort of propagate to the top of the type, and we'd report the result as an error. Unfortunately, this lost quite a lot of information about the context of the error.

The new approach is to treat the `Error` type as a wrapper *type function*, so if we encounter a broken type `t`, then we convert it to `Error "some message" t`.  This still bubbles up, but slightly differently, in that we should be preserving the whole type in the `t`. In particular, in the case of a constraint we should be saying `Error "some message" (fin inf)`.

I thought I updated all the places where these bubble up, but I guess I missed some.  The above probably happens when we are trying to make a constraint `fin inf` with `inf` being statically known, and then noticing that it would never work.  It looks like in some place we still have the behavior where the error is bubble up the old way.
